staging/src/k8s.io/client-go/discovery/restmapper.go

```diff
@@ -80,9 +80,20 @@ func NewRESTMapper(groupResources []*APIGroupResources, versionInterfaces meta.V
 				if !resource.Namespaced {
 					scope = meta.RESTScopeRoot
 				}
-				versionMapper.Add(gv.WithKind(resource.Kind), scope)
+
+				// this is for legacy resources and servers which don't list singular forms.  For those we must still guess.
+				if len(resource.SingularName) == 0 {
+					versionMapper.Add(gv.WithKind(resource.Kind), scope)
+					// TODO only do this if it supports listing
+					versionMapper.Add(gv.WithKind(resource.Kind+"List"), scope)
+					continue
+				}
+
+				plural := gv.WithResource(resource.Name)
+				singular := gv.WithResource(resource.SingularName)
+				versionMapper.AddSpecific(gv.WithKind(resource.Kind), plural, singular, scope)
 				// TODO only do this if it supports listing
-				versionMapper.Add(gv.WithKind(resource.Kind+"List"), scope)
+				versionMapper.AddSpecific(gv.WithKind(resource.Kind+"List"), plural, singular, scope)
```

Comment:
https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/v1beta1/types.go#L46

What if `ListKind` is not `Kind` + "List"? Looks like `APIResource` struct does not expose this information so we cannot use it here.